### DISCUSSION TITLE
Support for HTTP verbs on method

### DISCUSSION THREAD

**Discussion by colmexdev** [jeremyevans/forme#128] [General]
Hi there!

I went through the spec and code, but I didn't find something that could help me finding out how can form be provided with verbs rather than "get" or "post". That's why I come here to ask: Is that possible? If so, how can it be achieved?

I already enabled `all_verbs` plugin in Roda, so now I'm struggling on passing the verb through form method.

Thank you in advance!

**Comment by jeremyevans** [comment]
Browsers only support GET and POST.  Some web frameworks try to fake additional support using things like `Rack::MethodOverride`.  I do not recommend using verbs other than GET and POST if you are targeting browsers.  However, you can certainly use `Rack::MethodOverride` and include a `_method` parameter in your forms if you want to do that.

**Reply by colmexdev** [reply]
You're right, I think I got very used to The Rails Way&trade; that I was expecting for this to work automagically, even when it's not standard on browsers. Thank you very much for your advice!

**Reply by colmexdev** [reply]
One more thing for now. If I'd like to have a customized submit button (e.g., having an SVG inside of it). Is there a way to achieve this from a helper method? I think it can be donevia ERB templates, but it seems impossible from a helper method.

Thanks once again.

**Reply by jeremyevans** [reply]
I think a using a Forme::Formatter subclass that overrides the `format_submit` method as the :formatter option should do what you want.

**Reply by colmexdev** [reply]
I came to this solution:

```ruby
module Forme
  class Formatter::Submit < Formatter
    Forme.register_transformer(:formatter, :submit_delete, self)

    private

    def format_submit
      tag(:button, attr, [Forme::RawString.new(%Q[<svg height="20" width="20" aria-hidden="true" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 352 512"><path d="M242.7 256l100.1-100.1c12.3-12.3 12.3-32.2 0-44.5l-22.2-22.2c-12.3-12.3-32.2-12.3-44.5 0L176 189.3 75.9 89.2c-12.3-12.3-32.2-12.3-44.5 0L9.2 111.5c-12.3 12.3-12.3 32.2 0 44.5L109.3 256 9.2 356.1c-12.3 12.3-12.3 32.2 0 44.5l22.2 22.2c12.3 12.3 32.2 12.3 44.5 0L176 322.7l100.1 100.1c12.3 12.3 32.2 12.3 44.5 0l22.2-22.2c12.3-12.3 12.3-32.2 0-44.5L242.7 256z" fill="currentColor" /></svg>]), tag(:span, { class: "sr-only" }, "Remove")])
    end
  end
end
```
When applied as formatter, it parses the following:

```html
<button><svg height="20" width="20" aria-hidden="true" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 352 512"><path d="M242.7 256l100.1-100.1c12.3-12.3 12.3-32.2 0-44.5l-22.2-22.2c-12.3-12.3-32.2-12.3-44.5 0L176 189.3 75.9 89.2c-12.3-12.3-32.2-12.3-44.5 0L9.2 111.5c-12.3 12.3-12.3 32.2 0 44.5L109.3 256 9.2 356.1c-12.3 12.3-12.3 32.2 0 44.5l22.2 22.2c12.3 12.3 32.2 12.3 44.5 0L176 322.7l100.1 100.1c12.3 12.3 32.2 12.3 44.5 0l22.2-22.2c12.3-12.3 12.3-32.2 0-44.5L242.7 256z" fill="currentColor"></path></svg><span class="sr-only">Remove</span></button>
```

It works, but I'd kindly ask first isf it seems like a correct way to solve this.

**Reply by jeremyevans** [reply]
It does look kind of ugly with a raw string, but there isn't anything wrong it.  Using `tag(:svg, {height: 20, ...}, tag(:path, d: ...))` may be cleaner, but slower, as Forme has to convert everything itself.

Since you aren't interpolating, I would save the `Forme::RawString.new` call to a constant, and reference the constant inside `format_submit`.

**Reply by colmexdev** [reply]
I tried instantiating nested tags, but I couldn't make the `path` tag self-closing. I tried appending to the `SELF_CLOSING` constant in the default serializer and even subclassing de default serializer, but to no avail. Is there a way to expand self-closing tags' array?

**Reply by jeremyevans** [reply]
I would think that modifying `SELF_CLOSING` would work, but it isn't a good practice.  I should add a `self_closing_tag?(type)` method to the serializer to control whether the tag is self closing.

**Reply by jeremyevans** [reply]
I added `self_closing_tag?` in 0c12c489be28aa4b0f74ac9a844fb0e45f73d284
